src/scheduler.rs

```diff
@@ -0,0 +1,391 @@
+use bitcoin::consensus::Encodable;
+use bitcoin::psbt::PartiallySignedTransaction;
+use bitcoin::{Address, Script, TxOut};
+use ln_types::P2PAddress;
+use tonic_lnd::rpc::OpenChannelRequest;
+
+use crate::args::ArgError;
+use crate::lnd::{LndClient, LndError};
+
+#[derive(Clone, serde_derive::Deserialize)]
+pub struct ScheduledChannel {
+    pub(crate) node: P2PAddress,
+    #[serde(with = "bitcoin::util::amount::serde::as_sat")]
+    pub(crate) amount: bitcoin::Amount,
+}
+
+impl ScheduledChannel {
+    pub(crate) fn from_args(addr: &str, amount: &str) -> Result<Self, ArgError> {
+        let node = addr.parse::<P2PAddress>().map_err(ArgError::InvalidNodeAddress)?;
+
+        let amount = bitcoin::Amount::from_str_in(amount, bitcoin::Denomination::Satoshi)
+            .map_err(ArgError::InvalidBitcoinAmount)?;
+
+        Ok(Self { node, amount })
+    }
+}
+
+#[derive(Clone, serde_derive::Deserialize)]
+pub struct ScheduledPayJoin {
+    #[serde(with = "bitcoin::util::amount::serde::as_sat")]
+    wallet_amount: bitcoin::Amount,
+    pub(crate) channels: Vec<ScheduledChannel>,
+    pub(crate) fee_rate: u64,
+}
+
+impl ScheduledPayJoin {
+    pub(crate) fn new(
+        wallet_amount: bitcoin::Amount,
+        channels: Vec<ScheduledChannel>,
+        fee_rate: u64,
+    ) -> Self {
+        Self { wallet_amount, channels, fee_rate }
+    }
+
+    pub fn calculate_fees(&self) -> bitcoin::Amount {
+        let channel_count = self.channels.len() as u64;
+
+        let additional_vsize = if self.wallet_amount == bitcoin::Amount::ZERO {
+            (channel_count - 1) * (8 + 1 + 1 + 32) + 12
+        } else {
+            channel_count * (8 + 1 + 1 + 32)
+        };
+
+        bitcoin::Amount::from_sat(self.fee_rate * additional_vsize)
+    }
+
+    pub fn sum_channel_amounts(&self) -> bitcoin::Amount {
+        self.channels.iter().map(|chan| chan.amount).fold(bitcoin::Amount::ZERO, std::ops::Add::add)
+    }
+
+    pub fn wallet_amount(&self) -> bitcoin::Amount { self.wallet_amount }
+
+    /// Calculates the expected owned output value that this scheduled payjoin should have.
+    pub fn total_amount(&self) -> bitcoin::Amount {
+        let fees = self.calculate_fees();
+        let channel_amounts_sum = self.sum_channel_amounts();
+
+        fees + channel_amounts_sum + self.wallet_amount
+    }
+
+    /// Test connections with remote lightning nodes that we are trying to create channels with as
+    /// part of this [ScheduledPayJoin].
+    pub async fn test_connections(&self, client: &LndClient) -> Result<(), LndError> {
+        let handles = self
+            .channels
+            .iter()
+            .map(|ch| (client.clone(), ch.node.clone()))
+            .map(|(client, node)| {
+                tokio::spawn(async move { client.ensure_node_connected(node).await })
+            })
+            .collect::<Vec<_>>();
+
+        for handle in handles {
+            handle.await.unwrap()?;
+        }
+        Ok(())
+    }
```

Comment:
Testing connections is now done async.

Reply:
before each connection was done serially